gc/base/MemoryPoolAddressOrderedList.cpp

```diff
@@ -634,9 +700,29 @@ MM_MemoryPoolAddressOrderedList::internalAllocateTLH(MM_EnvironmentBase *env, ui
 	consumedSize = (maximumSizeInBytesRequired > freeEntrySize) ? freeEntrySize : maximumSizeInBytesRequired;
 	_largeObjectAllocateStats->decrementFreeEntrySizeClassStats(freeEntrySize);
 
-	/* If the leftover chunk is smaller than the minimum size, hand it out */
+	if (!alignTLHForGC(env, freeEntry, &consumedSize)) {
+		/* If alignment was required and it failed (i.e resulted in consumedSize < minEntrySize), abandon entry and retry */
+		abandonHeapChunk((void *)freeEntry, (void *)(((uint8_t *)freeEntry) + freeEntrySize));
+
+		_freeMemorySize -= freeEntrySize;
+		_allocDiscardedBytes += freeEntrySize;
+
+		entryNext = freeEntry->getNext(compressed);
+
+		if (entryNext == _firstUnalignedFreeEntry) {
+			_prevFirstUnalignedFreeEntry = FREE_ENTRY_END;
+		}
+
+		_heapFreeList = entryNext;
+		_freeEntryCount -= 1;
+
+		goto retry;
+	}
+
+	/* If the leftover chunk is smaller than the minimum size, hand it out. Handing out the leftover chunk may break alignment,
+	 * ensure we don't require alignment first. */
 	recycleEntrySize = freeEntrySize - consumedSize;
-	if (recycleEntrySize && (recycleEntrySize < _minimumFreeEntrySize)) {
+	if (recycleEntrySize && (recycleEntrySize < _minimumFreeEntrySize) && !extensions->isSATBBarrierActive()) {
```

Comment:
Introduce a helper isAlignementForParallelGCActive/Enabled/Required
that would just check: NULL != _tlhAlignmentBase

The same helper could be used in alignTLH mehtod, or even pulled out in front of the method (what would match the patter/pair of calls that we use for Card alignment)